### PR TITLE
feat(client): respect user color scheme in editor

### LINKED ISSUE #48936: Editor does not use the right theme
### Describe the Issue

The layout theme and the editor theme does not match for users who are not signed in.

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-1

### Steps to Reproduce

1. Open an incognito window on a browser 
2. Go to https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-1
3. Compare the layout theme and the editor theme

### Expected behavior

The editor should show the same theme as the layout.

### Screenshots

Screen Shot 2023-01-03 at 3 12 59 PM

### System

- Browser: [Chrome, Safari]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Sboonny**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #48936

### CODE REVIEW

**Review by raisedadead — COMMENTED** [review]
Thanks, can you also add a test because you are fixing a regression?

**Comment by Sboonny** [comment]
Nope, the test would require creating a cypress test, and the time and the effort it would take me to create a cypress test is not worth it, somebody who have full access to cypress tools can do a better job than me.

I want to focus on UI-Component issue.

File: client/src/templates/Challenges/classic/editor.tsx [hunk 1]
```diff
@@ -1203,8 +1203,16 @@ const Editor = (props: EditorProps): JSX.Element => {
...
-  const { theme } = props;
-  const editorTheme = theme === Themes.Night ? 'vs-dark-custom' : 'vs-custom';
+  const { isSignedIn, theme } = props;
+  const preferDarkScheme = window.matchMedia(
+    '(prefers-color-scheme: dark)'
+  ).matches;
+  const editorTheme =
+    theme === Themes.Night
+      ? 'vs-dark-custom'
+      : preferDarkScheme && !isSignedIn
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/classic/editor.tsx:1213]:
```suggestion
    theme === Themes.Night || (preferDarkScheme && !isSignedIn)
```
Nested ternaries are a pain to read and this is more concise.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

**Review by raisedadead — CHANGES_REQUESTED** [review]
One test is worth the time wasted in fixing a regression. Keeping this on hold for a quick test. 

Is anyone on @freeCodeCamp/dev-team willing to take it up? I might beat you to it over the week.

**Comment by ojeytonwilliams** [comment]
A Cypress test for this is probably overkill. All we really care about is the value of the `theme` prop that's passed to `MonacoEditor` in a few situations.

A jest test with a mocked `MonacoEditor` should be just fine.

**Comment by raisedadead** [comment]
Agreed — We want to include tests every time we introduce regressions.

**Comment by Sboonny** [comment]
Needed tests
---

1. When the user is logged in and have `preferred-color-scheme` set to `dark`, the editor theme shouldn't change to dark when they have picked default theme.
2. When the user isn't logged in and have `preferred-color-scheme` set to `dark`, the editor should change to dark.

**Comment by ojeytonwilliams** [comment]
Thinking about it, this bug only exists because the logic governing which theme is used is duplicated. It is defined in both editor.tsx and default.tsx, but should only be defined once.

**Comment by Sboonny** [comment]
Is there a way to set theme value for not logged in user, without local storage?

**Comment by ojeytonwilliams** [comment]
Only cookies, to my knowledge, and they're like local storage, but slightly worse (they're transmitted with the request, but there's nothing gained by doing so). Why do you ask?

**Comment by Sboonny** [comment]
The current state management ignores `theme` values, if the user isn't logged in.

**Comment by ojeytonwilliams** [comment]
Throwing local storage into the mix complicates things, but fortunately we aren't using it yet, so we don't need to worry about it.

Anyways, isn't https://github.com/freeCodeCamp/freeCodeCamp/pull/48757/ going to change this logic in a fairly fundamental way?

With that in mind @raisedadead how about we treat this as a hotfix and move the test requirement into https://github.com/freeCodeCamp/freeCodeCamp/pull/48757/

**Comment by raisedadead** [comment]
Sure, as long as it is not forgotten.

**Comment by Sboonny** [comment]
I will create issue and link system one to it

**Review by ojeytonwilliams — APPROVED** [review]
This all LGTM :+1:

File: client/src/templates/Challenges/classic/editor.tsx [hunk 2]
```diff
@@ -1203,8 +1203,14 @@ const Editor = (props: EditorProps): JSX.Element => {
...
-  const { theme } = props;
-  const editorTheme = theme === Themes.Night ? 'vs-dark-custom' : 'vs-custom';
+  const { isSignedIn, theme } = props;
+  const preferDarkScheme = window.matchMedia(
+    '(prefers-color-scheme: dark)'
+  ).matches;
+  const editorTheme =
+    theme === Themes.Night || (preferDarkScheme && !isSignedIn)
+      ? 'vs-dark-custom'
+      : 'vs-custom';
```
Comment (auchynnikau) [inline client/src/templates/Challenges/classic/editor.tsx:1213]: Let's increase condition readability a bit:
```suggestion
  const isDarkTheme = theme === Themes.Night || (preferDarkScheme && !isSignedIn);
  const editorTheme = isDarkTheme ? 'vs-dark-custom' : 'vs-custom';
```

**Review by raisedadead — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]

**Pull request merged**